### PR TITLE
[AMD][PyTorch] Redirect index_add_ to scatter_add_ on HIP for large tensors (#180430)

### PR DESCRIPTION
**Pull request by haoyuz**
Summary:

CONTEXT: On AMD MI350X, index_add_ with large source tensors and Zipf-distributed indices (common in MPZCH embedding tables) suffers from severe bf16 atomicAdd contention. The indexFuncLargeIndex kernel serializes on CAS-loop atomics when many threads target the same destination rows.

WHAT: When running on HIP with alpha=1, dim=0, and 2D tensors, redirect index_add_ to scatter_add_, which uses a different kernel path that avoids the atomicAdd contention bottleneck. The redirect now fires unconditionally for the matching shape; an earlier revision gated it on `source.size(0) > 100000`, but a size sweep showed scatter_add_ is at parity with index_add_ at n=1K and strictly faster from n=5K onward, so the threshold was unnecessarily conservative.

The redirect mirrors `index_add_cuda_impl`'s structured-kernel pre-copy: `if (!result.is_same(self)) result.copy_(self);` — required so that the out-of-place `torch.index_add(self, ...)` and `torch.index_add(self, ..., out=...)` paths (which share the same TORCH_IMPL_FUNC via structured_delegate) seed `result` before scatter_add_ accumulates into it.

Also adds:
- bench_index_add_powerlaw.py: reproduces the issue with Zipf-distributed indices at various alpha values, plus a size sweep that justifies removing the threshold.
- test_index_add_redirect_kernel_selection_hip in test_indexing.py: verifies via torch.profiler that the redirect actually fires for matching cases and is bypassed for alpha != 1, dim != 0, 1D, and 3D inputs (addressing reviewer feedback that prior tests only covered numerical correctness, not selection).
- test_index_add_correctness_2d_out_of_place: regression test for the structured-kernel pre-copy path (functional and explicit-out forms of torch.index_add).

Test Plan:
**Existing benchmark on AMD MI350X (gfx950), dst=[634364, 128] bf16, src=[9039274, 128] bf16:**

| Distribution     | index_add_ (before) | scatter_add_ (after) | Speedup |
|------------------|---------------------|----------------------|---------|
| Uniform          |          7,673 us   |          3,926 us    |   2.0x  |
| Zipf alpha=0.5   |          8,034 us   |          3,912 us    |   2.1x  |
| Zipf alpha=1.0   |         68,996 us   |         16,695 us    |   4.1x  |
| Zipf alpha=1.5   |        397,525 us   |         83,434 us    |   4.8x  |
| Zipf alpha=2.0   |        689,997 us   |        132,012 us    |   5.2x  |
| Zipf alpha=2.5   |        889,804 us   |        161,805 us    |   5.5x  |
| Zipf alpha=3.0   |      1,021,968 us   |        180,512 us    |   5.7x  |

Zipf alpha=1.5 matches the ~420,000 us observed in the IG Threads LSR Slimper training trace. The redirect achieves 4.8x speedup for this case.

**Threshold sweep on AMD MI350X (justifies removing the >100K gate). All bf16, dst=[634364, 128]:**

Uniform indices:

| src_size | index_add_ | scatter_add_ | Speedup |
|---------:|-----------:|-------------:|--------:|
|    1,000 |     21.6us |       21.5us |   1.00x |
|    5,000 |     24.5us |       21.5us |   1.14x |
|   10,000 |     28.4us |       22.6us |   1.26x |
|   50,000 |     62.0us |       36.8us |   1.69x |
|  100,000 |    105.0us |       55.5us |   1.89x |
|  200,000 |    193.7us |       93.3us |   2.08x |
|1,000,000 |    889.4us |      451.7us |   1.97x |
|9,039,274 |  7,698.5us |    3,935.3us |   1.96x |

Zipf alpha=1.5 (the contention regime that motivated this change):

| src_size | index_add_ | scatter_add_ | Speedup |
|---------:|-----------:|-------------:|--------:|
|    1,000 |     61.5us |       25.1us |   2.45x |
|    5,000 |    230.1us |       60.9us |   3.78x |
|   10,000 |    457.0us |      109.4us |   4.18x |
|   50,000 |  2,224.4us |      480.9us |   4.63x |
|  100,000 |  4,397.2us |      937.2us |   4.69x |
|9,039,274 |397,906.9us |   83,519.0us |   4.76x |

Conclusion: scatter_add_ is at parity at the smallest measured size and strictly faster from 5K onward, on every distribution tested. Removing the size gate is a clear win. Bench command:

```
buck run mode/dev-nosan-amd-gpu -m rocm70 -m fbcode//triton/cc:force_noop //scripts/haoyuz/amd/index_add:bench_index_add_powerlaw
```

**Numerical correctness on AMD MI350X (//scripts/haoyuz/amd/index_add:test_index_add_redirect, all 7 PASS):**

```
test_unique_indices_fp32              ... ok  (bit-exact)
test_unique_indices_bf16              ... ok  (no contention -> close to f32 ref)
test_small_exact_bf16                 ... ok  (small int inputs -> both paths exact in bf16)
test_duplicates_fp32_exact            ... ok  (deterministic atomicAdd in fp32)
test_contention_bf16_relative_error   ... ok  (L2 rel err < 0.01)
test_high_contention_zipf_bf16        ... ok  (L2 rel err < 0.5 even with 76K dupes)
test_redirect_kernel_selection_via_profiler ... ok  (profiler confirms scatter_add_ fires only for matching cases)
```

**test_indexing.py changes:**

- `test_index_add_correctness_2d_large` (fp32 + bf16 + fp16): Compares the redirected path to a CPU f32 reference for a 200K x 64 case with duplicate indices. Uses L2 relative error for low-precision dtypes (worst-case per-element error from ~200 atomic accumulations is too large for max-abs).
- `test_index_add_correctness_2d_out_of_place`: Regression test for the structured-kernel pre-copy. Exercises both `torch.index_add(self, ...)` (functional) and `torch.index_add(self, ..., out=...)` paths, which would silently produce garbage if the redirect skipped the `result.copy_(self)` seeding step.
- `test_index_add_correctness_2d_alpha`: Validates alpha != 1 path correctness (which always bypasses the redirect).
- `test_index_add_redirect_kernel_selection_hip`: HIP-only profiler-based test that asserts (a) redirect fires when alpha=1, dim=0, 2D, and (b) is bypassed for alpha != 1, dim != 0, 1D, and 3D inputs. This addresses reviewer feedback that previous tests only checked numerical correctness, not which kernel was selected.

**AI review fixes incorporated in this revision:**

- BUG (critical): Out-of-place `torch.index_add(self, ...)` and `torch.index_add(..., out=)` reach the same TORCH_IMPL_FUNC via `structured_delegate` and start with an uninitialized `result`. The redirect was missing the `result.copy_(self)` seeding that the legacy `index_add_cuda_impl` does at the top of its body — fixed by mirroring that pre-copy.
- Test gap: added `test_index_add_correctness_2d_out_of_place` regression for the above.
- Test fix: `test_small_exact_bf16` previously computed `dst_before` (bypass) but never asserted on it; added assertion.
- Test stability: `test_index_add_correctness_2d_large` used `atol=1e-1` per-element max for bf16, which is borderline tight for 200 dupes/row; switched to L2 relative error (< 1e-2).
- Test guard: standalone `test_index_add_redirect.py` is now class-level HIP-only (numerical comparisons on NVIDIA exercise the same kernel for both paths, providing no real coverage).

Co-authored-by: Claude

Reviewed By: echen4096

Differential Revision: D100635529

### CODE REVIEW

**Review by echen4096 — APPROVED** [review]

**Comment by jeffdaily** [comment]
@haoyuz did https://github.com/pytorch/pytorch/pull/172476 cause a regression?  Is that the real problem here?

**Comment by haoyuz** [comment]
> @haoyuz did #172476 cause a regression? Is that the real problem here?

@jeffdaily The benchmarks (before or after) were done without the PR included, so it should be independent from that PR.

**Comment by albanD** [comment]
@pytorchbot revert -m "The unstable job hid the failure. But you can see drci above for real cuda errors on index_add_" -c ignoredsignal

See https://hud.pytorch.org/pytorch/pytorch/commit/0fae863fd047d98bdac3d968ffb379bfe574d2a8#73578341595-box for details.

**Comment by jithunnair-amd** [comment]
@pytorchbot rebase 

Trying to rebase so `keep-going` takes effect in ROCm jobs on trunk.yml

**Comment by jeffdaily** [comment]
@haoyuz let's figure out how to get this landed but making sure ROCm CI is fully passing. Yes, we're in UNSTABLE state, so we can't rely on the CI automation to block the landing of this PR.  We can trigger ciflow/trunk manually without asking pytorchbot to merge and double-check the signal.

**Comment by ngimel** [comment]
Why is scatter_add free from atomicAdd contention problem? It still uses atomicAdd, if there are a lot of repeating indices it would still contend

**Comment by nWEIdia** [comment]
>> Initiating merge automatically since Phabricator Diff has merged

Please help verify GB200 unit test status for: 

Mismatched elements: 65 / 6400 (1.0%)
Greatest absolute difference: 0.22900390625 at index (86, 12) (up to 0.05 allowed)
Greatest relative difference: 1.779054880142212 at index (123, 7) (up to 0.01 allowed)

To execute this test, run the following from the base repo dir:
    python test/test_indexing.py TestIndexingCUDA.test_index_add_correctness_2d_out_of_place_cuda_bfloat16

File: test/test_indexing.py [hunk 1]
```diff
@@ -2126,6 +2127,155 @@ def _run_and_check(dst_shape, dim, num_idx, num_dest_dim, tag):
...
+        ref = self_.cpu().float().clone()
+        ref.index_add_(0, idx.cpu(), src.cpu().float())
+
+        # Functional out-of-place: returns a fresh tensor.
+        out_func = torch.index_add(self_, 0, idx, src)
+        # Out-explicit: caller-provided destination tensor.
+        out_explicit = torch.empty_like(self_)
+        torch.index_add(self_, 0, idx, src, out=out_explicit)
+
+        atol = 5e-2 if dtype == torch.bfloat16 else 1e-5
```
Comment (ngimel) [inline test/test_indexing.py:2196]: how did you set tolerances here? You have a lot of repeating indices (num_dest = 200, num_src=5000, so you are generating 5000 numbers in [0-200) range), and bf16 truncation for this will be bad, unless scatter_add on amd is actually diverted again into index_put implementation that does accumualtion in fp32. In any case, this is causing nvidia test failures which are expected due to rounding

Comment (haoyuz) [reply test/test_indexing.py:2196]: Thanks for pointing this out. Apologies that I missed the threshold check for NVIDIA -- will make sure it passes before reland.

**Comment by haoyuz** [comment]
> Why is scatter_add free from atomicAdd contention problem? It still uses atomicAdd, if there are a lot of repeating indices it would still contend

@ngimel  PR #146264 did the original optimization for `scatter_add`. Basically, `scatter_add` mitigates the global atomic contention by [using `opportunistic_fastAtomicAdd`](https://github.com/pytorch/pytorch/blob/3506be738d3fffcfc9359f96dfb43dd39965ea6c/aten/src/ATen/native/cuda/ScatterGatherKernel.cu#L42) which performs warp-level reduction before issuing the global atomics.
